crypto/x509/x509_vfy.c

```diff
@@ -362,6 +362,20 @@ int X509_verify_cert(X509_STORE_CTX *ctx) {
         ok = 0;
         goto end;
       }
+
+      // OpenSSL 1.1.1 continuously re-checks for trust and breaks the loop
+      // as soon as trust has been established. 1.0.2 builds the chain with all
+      // possible certs first and only checks for trust if the final cert in
+      // the chain is self-signed.
+      // This caused additional unanticipated certs to be in the established
+      // certificate chain, particularly when |X509_V_FLAG_PARTIAL_CHAIN| was
+      // set. We try checking continuously for trust here for better 1.1.1
+      // compatibility.
+      trust = check_trust(ctx);
+      if (trust == X509_TRUST_TRUSTED || trust == X509_TRUST_REJECTED) {
+        break;
+      }
```

Comment:
The tests also passes when this logic is conditioned on `X509_V_FLAG_PARTIAL_CHAIN`:
```
if (0 != (ctx->param->flags & X509_V_FLAG_PARTIAL_CHAIN)) {
  trust = check_trust(ctx);
  if (trust == X509_TRUST_TRUSTED || trust == X509_TRUST_REJECTED) {
    break;
  }
}
```
* Should there also be a test that confirms this behavioral change when `X509_V_FLAG_PARTIAL_CHAIN` has not been set? (Or do we only care about this behavior change when the flag is set?)

Reply:
* The real logic with `X509_V_FLAG_PARTIAL_CHAIN` is being done in `check_trust`, where we return for `X509_TRUST_TRUSTED` if it's set and we have a partial chain.
https://github.com/aws/aws-lc/blob/d4810f4f0530c9d9befbfe936c1b23b42faeb758/crypto/x509/x509_vfy.c#L770-L795

* I agree and we do have a test the checks that we have the intended behavioral change. The test above uses `for (bool partial_chain : {true, false}) {` to set the flag and then checks the found certificate chain length based on `partial_chain` with `    EXPECT_EQ(sk_X509_num(chain), partial_chain ? 2u : 3u);`.